proposals/pr002-hunkspace.md

```diff
@@ -0,0 +1,545 @@
+the file `//source/browser/components/preferences/preferences.xhtml`.
+
+[sfx]: https://searchfox.org/mozilla-central/source/browser/components/preferences/preferences.xhtml
+
+```html
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml"
+        xmlns:html="http://www.w3.org/1999/xhtml"
+        role="document"
+        id="preferences-root">
+<head>
+  <title data-l10n-id="pref-page-title"></title>
+  <link rel="localization" href="branding/brand.ftl"/>
+  <link rel="localization" href="browser/branding/brandings.ftl"/>
+  <!-- … -->
+</head>
+<!-- … -->
+<html:body/>
+```
+
+In turn, for example the file [`browser/branding/brandings.ftl`][brandings] has
+the following, with some details omitted for brevity:
+
+[brandings]: https://searchfox.org/mozilla-central/source/browser/locales/en-US/browser/branding/brandings.ftl
+
+```
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+-facebook-container-brand-name = Facebook Container
+-lockwise-brand-name = Firefox Lockwise
+-lockwise-brand-short-name = Lockwise
+-monitor-brand-name = Firefox Monitor
+-monitor-brand-short-name = Monitor
+-pocket-brand-name = Pocket
+-send-brand-name = Firefox Send
+-screenshots-brand-name = Firefox Screenshots
+```
+
+One will notice that the declaration of `preferences.xhtml` admits
+the UI author to declare the source for the bindings, i.e. the values
+for specific message identifiers, by specifying a path to a file that
+contains the bindings.
+
+This, in turn, means that the value to which a certain message
+identifier is bound, for example `-monitor-brand-name` (which can be
+seen in `brandings.ftl`), may depend based on the binding decision
+taken in the context of the UI widget.  This also means that another UI
+file may decide to use the same message identifier in a different UI
+widget context.
+
+While the decision to go to such lengths is at the discretion of the
+implementor, it nevertheless compels us to provide a way to address such
+a binding to the data provider.  That is, the data provider must be able
+to address a specific message.
+
+For this purpose, I propose using "context specifiers", which allow adding
+"auxiliary" data that allows disambiguation of the context.   Since the aim of
+the data provider is to form a flat key space, and the nature of the auxiliary
+data may have arbitrary internal structure, it compels us to move the auxiliary
+data outside of the "regular" key dimensions.  For this reason, the "auxiliary"
+data is specified as URI query parameters, or with JSON arguments that start
+with a "?" symbol.   The resulting URI and JSON forms are given below.
+
+URI:
+
+```
+mozilla-l10n://mozilla.org/app/firefox/locale/en-us/message-id/-monitor-brand-name⮒
+    ?from=browser/branding/brandings.ftl⮒
+    &to=source/browser/components/preferences/preferences.xhtml
+```
```

Comment:
I think this is a bit misaligned with how Fluent contexts, work. The response is not going to be different depending on `to`, but it will be different depending on the combination of resources that compose a given context.

```html
<head>
  <link rel="localization" href="toolkit/menubar.ftl>
  <link rel="localization" href="browser/privacy.ftl"/>
</head>
```

is a context, much like if you linked two JS scripts they'd be loaded into the same JS context.

So, for your use case, you'd either have to do:

```
mozilla-l10n://mozilla.org/app/firefox/locale/en-us/message-id/-monitor-brand-name⮒
    ?from=toolkit/menubar.ftl&from=browser/privacy.ftl
```

to indicate the context, or get a request with the list of URIs to construct a context, get its "guid" and then request for message IDs giving the context GUID.
Since the context is dynamic through the document's lifecycle, you'd also need to be able to "add" and "remove" resources from the context, much like you can with CSS style links during documents lifetime.

Reply:
Thanks for explaining. 

I was vaguely aware of this use case; and think that for a provider with a fully dynamic lifecycle the best we can do indeed is to request a context token (guid) from the provider, and then pass the token in the request.

This I believe is because we can not know whether the context is client-specific, or global, nor can we make any ex-ante assumptions about the current point in the provider's lifecycle without actually asking the provider.

I think that being able to declare a static context is always better when possible.  But of course, that may not be always possible, especially if the context needs to be retrofitted into existing behavior -- as is the case here.

Do you think the GUID approach would be acceptable in this case?

Reply:
I think it's a good start.

It's hard to predict for me how it'll work just yet and there will be some interesting dynamic around trying to get refcounting for the contexts shared between documents (or not - CoW contexts?), but I think it's a good start.

Reply:
It seems to me that the proposal here does not preclude the GUID approach -

Reply:
To represent Mozilla use case more realistically, you'd really need to do:

```json
{
  "scheme": "mozilla-l10n",
  "source": "mozilla.org",
  "app": "firefox",
  "locales": ["ru", "fr", "en-US"],
  "resource_ids": [
    "browser/preferences.ftl",
    "toolkit/menu.ftl",
    "branding/brands.ftl"
  ],
  "message_keys": [
    {"id": "prefs-document-title", "args": null},
    {"id": "prefs-open", "args": null},
    {"id": "prefs-cancel", "args": null},
    {"id": "prefs-header", "args": { "user": "John" }},
  ]
}
 ```

and then the response could be:

```json
{
  "messages": [
    {  "value": "Your Preferences", "attrs": [], "locale": "ru" },
    {  "value": "Open", "attrs": { "accesskey": "o" }, "locale": "ru" },
    {  "value": "Cancel", "attrs": { "accesskey": "c" }, "locale": "ru" },
    {  "value": null, "attrs": { "label": "Hello, John" }, "locale": "en-US" },
  ],
  "errors": [],
}
 ```

which alternatively could be encoded as a two step process:

Request for context id:

```json
{
  "scheme": "mozilla-l10n",
  "source": "mozilla.org",
  "app": "firefox",
  "locales": ["ru", "fr", "en-US"],
  "resource_ids": [
    "browser/preferences.ftl",
    "toolkit/menu.ftl",
    "branding/brands.ftl"
  ],
}
 ```

Response:

```json
{
  "context_id": "34jf2n#3j3n1",
  "errors": [],
}
```

Request for messages:

```json
{
  "scheme": "mozilla-l10n",
  "source": "mozilla.org",
  "app": "firefox",
  "context_id": "34jf2n#3j3n1",
  "message_keys": [
    {"id": "prefs-document-title", "args": null},
    {"id": "prefs-open", "args": null},
    {"id": "prefs-cancel", "args": null},
    {"id": "prefs-header", "args": { "user": "John" }},
  ]
}
```

and then the response could be:

```json
{
  "messages": [
    {  "value": "Your Preferences", "attrs": [], "locale": "ru" },
    {  "value": "Open", "attrs": { "accesskey": "o" }, "locale": "ru" },
    {  "value": "Cancel", "attrs": { "accesskey": "c" }, "locale": "ru" },
    {  "value": null, "attrs": { "label": "Hello, John" }, "locale": "en-US" },
  ],
  "errors": [],
}
 ```

And then we'll just need to figure out how to handle cache invalidation, context cleaning, and handling of a scenario where context is not available anymore.

But if you want to capture Mozilla's (and I expect close to how MessageFormat 2.0 / Web L10n API) will want to work, you'll need more of that.
I don't expect a runtime API on critical path of first paint to send 900 requests for IDs, rather a single request with 900 IDs.